### PR TITLE
Synched animations

### PR DESCRIPTION
**Pull request by davidmanzanares**
Fix https://github.com/CartoDB/carto-vl/issues/599

### CODE REVIEW

File: debug/multi-anim.html [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            center: [-3.589442037237177, 40.614912811223235],
+            zoom: 15,
+            dragRotate: false
+        });
+        carto.setDefaultAuth({
+            user: 'rochoa',
+            apiKey: 'default_public'
+        });
+        const s = carto.expressions;
+        const jaramaSource = new carto.source.Dataset('circuito_del_jarama');
```
Comment (jgoizueta) [inline debug/multi-anim.html]: This dataset is very nice for examples, maybe we should ask @rochoa permission to put it in cartovl 😛

File: debug/multi-anim.html [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        const racetrackLayer = new carto.Layer('racetrack', jaramaSource, new carto.Viz(`
```
Comment (jgoizueta) [inline debug/multi-anim.html]: I mean, this one too, of course

File: src/core/viz/expressions/Animation.js [hunk 3]
```diff
@@ -93,7 +96,46 @@ export class Animation extends BaseExpression {
...
+                    return true;
+                }
+                return false;
+            }));
+            if (waitingForLayer.size > 0) {
+                return;
+            }
+            [...waitingForOthers.values()].map(anim => {
+                if (anim._paused === 'default') {
+                    anim._paused = false;
```
Comment (jgoizueta) [inline src/core/viz/expressions/Animation.js]: should be anim.play() right?

File: src/core/viz/expressions/Animation.js [hunk 4]
```diff
@@ -93,7 +96,46 @@ export class Animation extends BaseExpression {
         return !this.paused;
     }
 
+    _dataReady() {
```
Comment (jgoizueta) [inline src/core/viz/expressions/Animation.js]: It takes some time to understand this logic 😅 and I'm afraid it'll take some time each time, but it seems right. I think I'd break it into some functions to try to make clear the intent of each piece, but maybe that could end up worse... so do as you please

File: src/core/viz/expressions/Animation.js [hunk 5]
```diff
@@ -93,7 +96,46 @@ export class Animation extends BaseExpression {
...
+                    anim._paused = false;
+                }
+            });
+            waitingForOthers.clear();
+        }
+    }
+
+    _postShaderCompile(program, gl) {
+        waitingForLayer.add(this);
+        this._paused = 'default';
```
Comment (jgoizueta) [inline src/core/viz/expressions/Animation.js]: I don't find `default`quite clear... maybe `'uninitialized'`? `'waiting-for-start'`? `'waiting'`?

File: src/core/viz/expressions/Animation.js [hunk 6]
```diff
@@ -93,7 +96,46 @@ export class Animation extends BaseExpression {
...
+        super._postShaderCompile(program, gl);
+    }
+
     _setTimestamp(timestamp) {
+        super._setTimestamp(timestamp);
+
+        if (this._paused && this._lastTime === undefined) {
```
Comment (jgoizueta) [inline src/core/viz/expressions/Animation.js]: This should warrant some comment for the future... although I understand anything you write could be potentially more puzzling

File: src/core/viz/expressions/Animation.js (hunk 5 above)
Comment (elenatorro) [reply src/core/viz/expressions/Animation.js]: Which are the values could have `paused`? `true`, `false` and 'default? Maybe, instead, I would use a different property `this._initialized` instead, but I have to take a deeper look.

**Comment by elenatorro** [comment]
After reading [this comment](https://github.com/CartoDB/carto-vl/issues/599#issuecomment-400707023) where we discuss what happens if the user wants to synchronize animations, what if we add a method to do that? So if the user has

`viz1.animation`, `viz2.animation` and `viz3.animation` 

he can:

`viz2.animation.sync([viz2.animation, viz3.animation]);` 

What do you think?

File: src/core/viz/expressions/Animation.js [hunk 7]
```diff
@@ -6,6 +6,9 @@ import Property from './basic/property';
 import Variable from './basic/variable';
 import { castDate } from '../../../api/util';
 
+let waitingForLayer = new Set();
```
Comment (IagoLast) [inline src/core/viz/expressions/Animation.js]: const?

File: debug/multi-anim.html (hunk 1 above)
Comment (davidmanzanares) [reply debug/multi-anim.html]: Pinging @rochoa :smile:

File: src/core/viz/expressions/Animation.js (hunk 7 above)
Comment (davidmanzanares) [reply src/core/viz/expressions/Animation.js]: Nop, it is reset on line 105

**Comment by davidmanzanares** [comment]
@elenatorro What would be the semantics of `viz1.animation.sync([viz2.animation, viz3.animation]);`?

At first, I had assumed that `viz1` would be set to the time/progress of the parameter, but then I saw that actually, it must be the other way since you have multiple animations in the array.

For me, it's strange to call a method with the other animations and to actually change the other two, while the first one is left as-is.

I think we could leave this to another PR.

**Comment by davidmanzanares** [comment]
So... approved?

File: src/renderer/viz/expressions/Animation.js [hunk 8]
```diff
@@ -93,7 +96,46 @@ export class Animation extends BaseExpression {
         return !this.paused;
     }
 
+    _dataReady() {
```
[the remark this answers is not in the document]
Comment (davidmanzanares) [reply src/renderer/viz/expressions/Animation.js]: It is not API public, and this is consistent with other methods in the expressions, setting and complying with a new, to be discussed, rule is outside the scope of this

**Pull request merged**